h/api/models/document.py

```diff
@@ -182,6 +186,153 @@ def __repr__(self):
         return '<DocumentMeta %s>' % self.id
 
 
+def create_or_update_document_uri(db,
+                                  claimant,
+                                  uri,
+                                  type,
+                                  content_type,
+                                  document,
+                                  created,
+                                  updated):
+    """
+    Create or update a DocumentURI with the given parameters.
+
+    If an equivalent DocumentURI already exists in the database then its
+    updated time will be updated.
+
+    If no equivalent DocumentURI exists in the database then a new one will be
+    created and added to the database.
+
+    To be considered "equivalent" an existing DocumentURI must have the same
+    claimant, uri, type and content_type, but the Document object that it
+    belongs to may be different. The claimant and uri are normalized before
+    comparing.
+
+    :param db: the database session
+    :type db: sqlalchemy.orm.session.Session
+
+    :param claimant: the .claimant property of the DocumentURI
+    :type claimant: unicode
+
+    :param uri: the .uri property of the DocumentURI
+    :type uri: unicode
+
+    :param type: the .type property of the DocumentURI
+    :type type: unicode
+
+    :param content_type: the .content_type property of the DocumentURI
+    :type content_type: unicode
+
+    :param document: the Document that the new DocumentURI will belong to, if a
+        new DocumentURI is created
+    :type document: h.api.models.Document
+
+    :param created: the time that will be used as the .created time for the new
+        DocumentURI, if a new one is created
+    :type created: datetime.datetime
+
+    :param updated: the time that will be set as the .updated time for the new
+        or existing DocumentURI
+    :type updated: datetime.datetime
+
+    """
+    docuri = DocumentURI.query.filter(
+        DocumentURI.claimant_normalized == normalize(claimant),
+        DocumentURI.uri_normalized == normalize(uri),
+        DocumentURI.type == type,
+        DocumentURI.content_type == content_type).first()
+
+    if docuri is None:
+        docuri = DocumentURI(claimant=claimant,
+                             uri=uri,
+                             type=type,
+                             content_type=content_type,
+                             document=document,
+                             created=created,
+                             updated=updated)
+        db.add(docuri)
+    elif not docuri.document == document:
+        log.warn('Found DocumentURI with id %d does not match expected '
+                 'document with id %d', docuri.document_id, document.id)
```

Comment:
Looks like there is a mismatch the interpolated string and the passed in variables (DocumentURI id gets replaced with `docuri.document_id`). In any case, I think we should add all three ids: `docuri.id`, `docuri.document_id`, `document.id`.